Userland/Applications/Calendar/CalendarWidget.cpp

```diff
@@ -0,0 +1,331 @@
+        unsigned view_month = m_calendar_frame->view_month();
+        unsigned view_year = m_calendar_frame->view_year();
+        if (m_calendar_frame->mode() == GUI::Calendar::Month) {
+            view_month++;
+            if (m_calendar_frame->view_month() == 12) {
+                view_month = 1;
+                view_year++;
+            }
+        } else {
+            view_year++;
+        }
+        m_calendar_frame->update_tiles(view_year, view_month);
+    });
+}
+
+void CalendarWidget::update_window_title()
+{
+    StringBuilder builder;
+    if (current_filename().is_empty())
+        builder.append("Untitled"sv);
+    else
+        builder.append(current_filename());
+    builder.append("[*] - Calendar"sv);
+
+    window()->set_title(builder.to_deprecated_string());
+}
+
+ErrorOr<NonnullRefPtr<GUI::Action>> CalendarWidget::create_add_event_action()
+{
+    return GUI::Action::create("&Add Event", {}, TRY(Gfx::Bitmap::load_from_file("/res/icons/16x16/add-event.png"sv)), [&](const GUI::Action&) {
+        AddEventDialog::show(m_calendar_frame->selected_date(), *m_calendar, window());
+    });
+}
+
+ErrorOr<NonnullRefPtr<GUI::Action>> CalendarWidget::create_jump_to_action()
+{
+    return GUI::Action::create("Jump to &Today", {}, TRY(Gfx::Bitmap::load_from_file("/res/icons/16x16/calendar-date.png"sv)), [&](const GUI::Action&) {
+        m_calendar_frame->set_selected_date(Core::DateTime::now());
+        m_calendar_frame->update_tiles(Core::DateTime::now().year(), Core::DateTime::now().month());
+    });
+}
+
+ErrorOr<NonnullRefPtr<GUI::Action>> CalendarWidget::create_view_month_action()
+{
+    return GUI::Action::create_checkable("&Month View", { Mod_Ctrl, KeyCode::Key_1 }, TRY(Gfx::Bitmap::load_from_file("/res/icons/16x16/calendar-month-view.png"sv)), [&](const GUI::Action&) {
+        if (m_calendar_frame->mode() == GUI::Calendar::Year)
+            m_calendar_frame->toggle_mode();
+    });
+}
+
+ErrorOr<NonnullRefPtr<GUI::Action>> CalendarWidget::create_view_year_action()
+{
+    return GUI::Action::create_checkable("&Year View", { Mod_Ctrl, KeyCode::Key_2 }, TRY(Gfx::Bitmap::load_from_file("/res/icons/16x16/icon-view.png"sv)), [&](const GUI::Action&) {
+        if (m_calendar_frame->mode() == GUI::Calendar::Month)
+            m_calendar_frame->toggle_mode();
+    });
+}
+
+ErrorOr<NonnullRefPtr<GUI::Action>> CalendarWidget::create_open_settings_action()
+{
+    return GUI::Action::create("&Settings", {}, TRY(Gfx::Bitmap::load_from_file("/res/icons/16x16/app-settings.png"sv)), [&](GUI::Action const&) {
+        GUI::Process::spawn_or_show_error(window(), "/bin/CalendarSettings"sv);
+    });
+}
+
+void CalendarWidget::create_on_tile_doubleclick()
+{
+    m_calendar_frame->on_tile_doubleclick = [&] {
+        AddEventDialog::show(m_calendar_frame->selected_date(), *m_calendar, window());
+    };
+}
+
+void CalendarWidget::create_on_tile_paint()
+{
+    m_calendar_frame->on_tile_paint = [&](GUI::Calendar::Tile& tile) {
```

Comment:
This factoring feels a bit out of place compared to what we normally do with LibGUI widgets.

Instead of having an `on_tile_paint` hook, I suggest allowing custom tile painting via subclassing.

If `GUI::Calendar` has a virtual `paint_tile()` function, we can subclass & override it in the Calendar application. :)

Reply:
Good to know, I created a subclass `Calendar::EventCalendar` that inherits from `GUI::Calendar`. I added a `paint_tile` method to `GUI::Calendar` that is called for each tile and is used by `Calendar::EventCalendar`.

Reply:
@monroeclinton The first commit adds the on_tile_paint hook and the second one removes it. Please fixup squash the iterations rather than keeping the history of your journey to a better design in the commits.

Reply:
@ADKaster Fixed